sympy/matrices/reductions.py

```diff
@@ -334,14 +330,9 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
     if you depend on the form row reduction algorithm leaves entries
     of the matrix, set ``noramlize_last=False``
     """
-
-    if M.is_square and _check_expr(M):
-        domain = _eval_domain(M)
-        extension = None
-        if domain is True:
-            extension = True
-
-        DOM = DomainMatrix.from_Matrix(M, extension=extension, field=True)
+    domain = _eval_domain(M)
+    if M.is_square and _check_expr(M) and domain != EX:
+        DOM = DomainMatrix.from_Matrix(M, radicals_limit=1, field=True)
```

Comment:
Since we have already evaluated the domain using `_eval_domain`, we should directly be creating DomainMatrix as `DomainMatrix.__init__`.
```py
DomainMatrix(M.tolist(), M.shape, domain=domain)
```
But this causes us issues as:
On using `DomainMatrix.__init__`, the DOM is 
```py
DomainMatrix([[70 + 69*sqrt(2), 22*sqrt(2) + 41, 84 + 72*sqrt(2)], [12*sqrt(2) + 23, 12 + 49*sqrt(2), 5*sqrt(2) + 78], [27 + 67*sqrt(2), 37 + 49*sqrt(2), 33*sqrt(2) + 50]], (3, 3), QQ<sqrt(2)>)
```
On using the `DomainMatrix.from_Matrix`, the DOM is 
```py
DomainMatrix([[ANP([69, 70], [1, 0, -2], QQ), ANP([22, 41], [1, 0, -2], QQ), ANP([72, 84], [1, 0, -2], QQ)], [ANP([12, 23], [1, 0, -2], QQ), ANP([49, 12], [1, 0, -2], QQ), ANP([5, 78], [1, 0, -2], QQ)], [ANP([67, 27], [1, 0, -2], QQ), ANP([49, 37], [1, 0, -2], QQ), ANP([33, 50], [1, 0, -2], QQ)]], (3, 3), QQ<sqrt(2)>)
```
This difference is due to the fact that `_sympify` runs on every element via `DomainMatrix.from_list_sympy`.

@oscarbenjamin, kindly share your opinion.
Thanks.

Reply:
This diff seems to work for all domains i.e. including `EX` and for the above example also:
```py
diff --git a/sympy/matrices/reductions.py b/sympy/matrices/reductions.py
index 69f68951c0..30fae60aeb 100644
--- a/sympy/matrices/reductions.py
+++ b/sympy/matrices/reductions.py
@@ -5,6 +5,7 @@
 from sympy.polys.matrices.domainmatrix import DomainMatrix
 from sympy.core import Expr, Pow
 from sympy.polys.domains import EX
+from sympy.core.sympify import _sympify

 from .utilities import _get_intermediate_simp, _iszero
 from .determinant import _find_reasonable_pivot
@@ -269,11 +270,12 @@ def _check_expr(M):
     return False


-def _eval_domain(M):
-    flat_list = [item for sublist in M.tolist() for item in sublist]
+def _eval_domain(M, rows):
+    items_sympy = [_sympify(item) for row in M.tolist() for item in row]
     primitive_element = {p for p in M.atoms(Pow) if p.is_number and p.exp.is_Rational and p.exp.q != 1}
-    domain = DomainMatrix.get_domain(flat_list, radicals_limit=len(primitive_element), field=True)[0]
-    return domain
+    domain, items_domain = DomainMatrix.get_domain(items_sympy, radicals_limit=len(primitive_element), field=True)
+    domain_rows = [[items_domain[rows*r + c] for c in range(rows)] for r in range(rows)]
+    return domain, domain_rows


 def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
@@ -330,9 +332,11 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
     if you depend on the form row reduction algorithm leaves entries
     of the matrix, set ``noramlize_last=False``
     """
-    domain = _eval_domain(M)
-    if M.is_square and _check_expr(M) and domain != EX:
-        DOM = DomainMatrix.from_Matrix(M, radicals_limit=1, field=True)
+
+    if M.is_square and _check_expr(M):
+        rows = M.shape[0]
+        domain, items_domain = _eval_domain(M, rows)
+        DOM = DomainMatrix(items_domain, (rows, rows), domain=domain)
         DOM_mat, pivot_cols = DOM.rref()
         mat = DOM_mat.to_Matrix()
     else:
(END)   
```
But now, `_eval_domain` is same as `DomainMatrix.from_list_sympy`, perhaps, calling `DomainMatrix.form_Matrix` is better.

Reply:
`DomainMatrix.__init__` expects the input to be already in the domain so you'd need to convert them all using `domain.from_sympy`. Maybe `from_Matrix` and `from_list_sympy` should be able to take a `domain` argument to say which domain it should go to.

Reply:
> This diff seems to work for all domains but excluding `EX` and for the above example also:
> 
> ```python
> diff --git a/sympy/matrices/reductions.py b/sympy/matrices/reductions.py
> index 69f68951c0..30fae60aeb 100644
> --- a/sympy/matrices/reductions.py
> +++ b/sympy/matrices/reductions.py
> @@ -5,6 +5,7 @@
>  from sympy.polys.matrices.domainmatrix import DomainMatrix
>  from sympy.core import Expr, Pow
>  from sympy.polys.domains import EX
> +from sympy.core.sympify import _sympify
> 
>  from .utilities import _get_intermediate_simp, _iszero
>  from .determinant import _find_reasonable_pivot
> @@ -269,11 +270,12 @@ def _check_expr(M):
>      return False
> 
> 
> -def _eval_domain(M):
> -    flat_list = [item for sublist in M.tolist() for item in sublist]
> +def _eval_domain(M, rows):
> +    items_sympy = [_sympify(item) for row in M.tolist() for item in row]
>      primitive_element = {p for p in M.atoms(Pow) if p.is_number and p.exp.is_Rational and p.exp.q != 1}
> -    domain = DomainMatrix.get_domain(flat_list, radicals_limit=len(primitive_element), field=True)[0]
> -    return domain
> +    domain, items_domain = DomainMatrix.get_domain(items_sympy, radicals_limit=len(primitive_element), field=True)
> +    domain_rows = [[items_domain[rows*r + c] for c in range(rows)] for r in range(rows)]
> +    return domain, domain_rows
> 
> 
>  def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
> @@ -330,9 +332,11 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
>      if you depend on the form row reduction algorithm leaves entries
>      of the matrix, set ``noramlize_last=False``
>      """
> -    domain = _eval_domain(M)
> -    if M.is_square and _check_expr(M) and domain != EX:
> -        DOM = DomainMatrix.from_Matrix(M, radicals_limit=1, field=True)
> +
> +    if M.is_square and _check_expr(M):
> +        rows = M.shape[0]
> +        domain, items_domain = _eval_domain(M, rows)
> +        DOM = DomainMatrix(items_domain, (rows, rows), domain=domain)
>          DOM_mat, pivot_cols = DOM.rref()
>          mat = DOM_mat.to_Matrix()
>      else:
> (END)   
> ```
> 
> But now, `_eval_domain` is same as `DomainMatrix.from_list_sympy`, perhaps, calling `DomainMatrix.form_Matrix` is better.

Reply:
@oscarbenjamin, but it doesn't work as from_list_sympy calls `get_domain`:
```py
In [1]: from sympy.polys.matrices import DomainMatrix

In [2]: A=Matrix([[70 + 69*sqrt(2), 22*sqrt(2) + 41, 84 + 72*sqrt(2)], [12*sqrt(2) + 23, 12 + 49*sqrt(2), 5*sqrt(2) + 7
   ...: 8], [27 + 67*sqrt(2), 37 + 49*sqrt(2), 33*sqrt(3) + 50]])

In [3]: dM=DomainMatrix.from_Matrix(A, domain='QQ<sqrt(2)>')

In [4]: dM
Out[4]: DomainMatrix([[EX(70 + 69*sqrt(2)), EX(22*sqrt(2) + 41), EX(84 + 72*sqrt(2))], [EX(12*sqrt(2) + 23), EX(12 + 49*sqrt(2)), EX(5*sqrt(2) + 78)], [EX(27 + 67*sqrt(2)), EX(37 + 49*sqrt(2)), EX(50 + 33*sqrt(3))]], (3, 3), EX)

In [5]: dM.domain
Out[5]: EX
```
Maybe if you suggest, we can make changes to `from_list_sympy` to consider the domain if provided as argument.

Secondly, do we need `_eval_domain`?
The below comment(without using `_eval_domain`) seems to work the same as these diffs(using `_eval_domain` i.e. predetermining the domain).
Thanks.

Reply:
The `from_list_sympy` function could be changed so that if a domain is given explicitly then it uses that instead of calling `get_domain`.

I'm not sure if we do need `_eval_domain`. There will be some cases where we will want to choose the domain carefully and I'm not sure that `construct_domain` can be used for that. Perhaps this extra logic can go into the `get_domain` function in general though.

Reply:
> The `from_list_sympy` function could be changed so that if a domain is given explicitly then it uses that instead of calling `get_domain`.
> 
`from_list_sympy`  will always need to call `get_domain` as it returns `items_domain`. Thus, I am unsure if we need `domain` option for it.
> Perhaps this extra logic can go into the `get_domain` function in general though.
>
Sure, will do so.